File: ps/leetcode_java/src/FlattenAMultilevelDoublyLinkedList/FlattenAMultilevelDoublyLinkedList.java
package FlattenAMultilevelDoublyLinkedList;

public class FlattenAMultilevelDoublyLinkedList {
    class Node {
        public int val;
        public Node prev;
        public Node next;
        public Node child;
    };
    // 처음 풀이
    public Node dfs(Node cur){
        Node nextNode = cur.next;
        if(cur.child != null){
            cur.next = cur.child;
            cur.child.prev = cur;
            Node res = dfs(cur.child);
            cur.child = null;
            if(nextNode!=null) res.next = nextNode;
            if(nextNode!=null) nextNode.prev = res;
        }
        if(nextNode == null){
            return cur;
        }
        else{
            return dfs(nextNode);
        }
    }
    public Node flatten(Node head) {
        if(head!=null) dfs(head);
        return head;
    }
}

File: ps/leetcode_cpp/FlattenAMultilevelDoublyLinkedList.cpp
#include <iostream>
#include <stack>
#include <queue>
#include <map>
#include <unordered_map>
using namespace std;

class Node {
public:
   int val;
   Node* prev;
   Node* next;
   Node* child;
};


class Solution {
public:
	// tail 노드를 return
	Node* chk(Node *iter) {
		if (iter->child) {
			Node *add = iter->next; // 4
			iter->child->prev = iter; // 7->3
			iter->next = iter->child; // 3->7
			iter->child = NULL;

			Node *tail = chk(iter->next); // chk(7) = 10
			if(add!=NULL){
				tail->next = add;
				add->prev = tail;
				return chk(add);
			}
			return tail;
		}
		else{
			if(iter->next==NULL) return iter;
			return chk(iter->next);
		}
	}

   Node* flatten(Node* head) {
	   	if(head==NULL) return NULL;
		Node *dummy = new Node();
		dummy->next = head;
		chk(head);
		return dummy->next;
   }
};